File: Collections/ComplementCollection.vb
Imports System.Collections.Specialized
Imports Ks.Common

Public Class ComplementingCollectionMaster(Of T)
    Implements INotifyingCollection(Of T)

    Public Sub New()
        Me._Collection1 = New ChildComplementCollection(Of T)(Me)
        Me._Collection2 = New ChildComplementCollection(Of T)(Me)
    End Sub

#Region "Collection1 Property"
    Private ReadOnly _Collection1 As ChildComplementCollection(Of T)

    Public ReadOnly Property Collection1 As INotifyingCollection(Of T)
        Get
            Return Me._Collection1
        End Get
    End Property
#End Region

#Region "Collection2 Property"
    Private ReadOnly _Collection2 As ChildComplementCollection(Of T)

    Public ReadOnly Property Collection2 As INotifyingCollection(Of T)
        Get
            Return Me._Collection2
        End Get
    End Property
#End Region

#Region "CollectionChanged Event"
    Private Event NonGeneric_CollectionChanged As NotifyCollectionChangedEventHandler Implements INotifyCollectionChanged.CollectionChanged
    Public Event CollectionChanged As NotifyCollectionChangedEventHandler(Of T) Implements INotifyCollectionChanged(Of T).CollectionChanged

    Protected Sub OnCollectionChanged(ByVal E As NotifyCollectionChangedEventArgs(Of T))
        RaiseEvent CollectionChanged(Me, E)
        RaiseEvent NonGeneric_CollectionChanged(Me, E)
    End Sub
#End Region

#Region "Reading Logic"
    Public ReadOnly Property Count As Integer Implements ICollection(Of T).Count
        Get
            Return Me._Collection1.Count + Me._Collection2.Count
        End Get
    End Property

    Public ReadOnly Property IsReadOnly As Boolean Implements ICollection(Of T).IsReadOnly
        Get
            Return False
        End Get
    End Property

    Public Sub CopyTo(array() As T, arrayIndex As Integer) Implements ICollection(Of T).CopyTo
        If array.Length - arrayIndex < Me.Count Then
            Throw New ArgumentException("The number of elements in the source System.Collections.Generic.ICollection`1 is greater than the available space from arrayIndex to the end of the destination array.")
        End If
        Me._Collection1.CopyTo(array, arrayIndex)
        Me._Collection2.CopyTo(array, arrayIndex + Me._Collection1.Count)
    End Sub

    Public Function Contains(item As T) As Boolean Implements ICollection(Of T).Contains
        Return Me._Collection1.Contains(item) OrElse
               Me._Collection2.Contains(item)
    End Function

    Public Function GetEnumerator() As IEnumerator(Of T) Implements IEnumerable(Of T).GetEnumerator
        Return Me._Collection1.Concat(Me._Collection2).GetEnumerator()
    End Function

    Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
        Return Me.GetEnumerator()
    End Function
#End Region

#Region "ChildOperation Logic"
    Friend Sub AddC(ByVal Item As T, ByVal C1 As ChildComplementCollection(Of T), Optional C2 As ChildComplementCollection(Of T) = Nothing)
        If C2 Is Nothing Then
            If C1 Is Me._Collection1 Then
                C2 = Me._Collection2
            Else
                C2 = Me._Collection1
            End If
        End If

        If Not C2.RemoveI(Item) Then
            Throw New ArgumentException()
        End If
        C1.AddI(Item)

        C1.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateAdd(Item))
        C2.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateRemove(Item))
    End Sub

    Friend Sub ClearC(ByVal C1 As ChildComplementCollection(Of T), Optional C2 As ChildComplementCollection(Of T) = Nothing)
        Dim C1Clone As IList(Of T)

        If C2 Is Nothing Then
            If C1 Is Me._Collection1 Then
                C2 = Me._Collection2
            Else
                C2 = Me._Collection1
            End If
        End If

        C1Clone = New T(C1.Count - 1) {}
        C1.CopyTo(C1Clone)

        C1.ClearI()

        For Each I As T In C1Clone
            C2.AddI(I)
            C2.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateAdd(I))
        Next

        C1.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateReset())
    End Sub

    Friend Function RemoveC(ByVal Item As T, ByVal C1 As ChildComplementCollection(Of T), Optional C2 As ChildComplementCollection(Of T) = Nothing) As Boolean
        If C2 Is Nothing Then
            If C1 Is Me._Collection1 Then
                C2 = Me._Collection2
            Else
                C2 = Me._Collection1
            End If
        End If

        If Not C1.RemoveI(Item) Then
            Throw New ArgumentException()
        End If
        C2.AddI(Item)

        C1.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateRemove(Item))
        C2.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateAdd(Item))

        Return True
    End Function
#End Region

    Public Sub Add(item As T) Implements ICollection(Of T).Add
        If Me._Collection1.Contains(item) OrElse Me._Collection2.Contains(item) Then
            Exit Sub
        End If

        Me._Collection1.Add(item)
        Me.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateAdd(item))
    End Sub

    Public Sub Clear() Implements ICollection(Of T).Clear
        Me._Collection1.Clear()
        Me._Collection2.Clear()
        Me.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateReset())
    End Sub

    Public Function Remove(item As T) As Boolean Implements ICollection(Of T).Remove
        If Me._Collection1.Remove(item) OrElse
           Me._Collection2.Remove(item) Then
            Me.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateRemove(item))
            Return True
        End If
        Return False
    End Function

End Class

Friend Class ChildComplementCollection(Of T)
    Implements INotifyingCollection(Of T)

    Friend Sub New(ByVal Master As ComplementingCollectionMaster(Of T))
        Me.InnerList = New List(Of T)()
        Me._Master = Master
    End Sub

    Private ReadOnly InnerList As List(Of T)

#Region "CollectionChanged Event"
    Private Event NonGeneric_CollectionChanged As NotifyCollectionChangedEventHandler Implements INotifyCollectionChanged.CollectionChanged
    Public Event CollectionChanged As NotifyCollectionChangedEventHandler(Of T) Implements INotifyCollectionChanged(Of T).CollectionChanged

    Protected Friend Sub OnCollectionChanged(ByVal E As NotifyCollectionChangedEventArgs(Of T))
        RaiseEvent CollectionChanged(Me, E)
        RaiseEvent NonGeneric_CollectionChanged(Me, E)
    End Sub
#End Region

#Region "Master Property"
    Private ReadOnly _Master As ComplementingCollectionMaster(Of T)

    Public ReadOnly Property Master As ComplementingCollectionMaster(Of T)
        Get
            Return Me._Master
        End Get
    End Property
#End Region

#Region "Reading Logic"
    Public ReadOnly Property Count As Integer Implements ICollection(Of T).Count
        Get
            Return Me.InnerList.Count
        End Get
    End Property

    Public ReadOnly Property IsReadOnly As Boolean Implements ICollection(Of T).IsReadOnly
        Get
            Return False
        End Get
    End Property

    Public Sub CopyTo(array() As T, arrayIndex As Integer) Implements ICollection(Of T).CopyTo
        Me.InnerList.CopyTo(array, arrayIndex)
    End Sub

    Public Function Contains(item As T) As Boolean Implements ICollection(Of T).Contains
        Return Me.InnerList.Contains(item)
    End Function

    Public Function GetEnumerator() As List(Of T).Enumerator
        Return Me.InnerList.GetEnumerator()
    End Function

    Private Function GetEnumerator_Interface() As IEnumerator(Of T) Implements IEnumerable(Of T).GetEnumerator
        Return Me.InnerList.GetEnumerator()
    End Function

    Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
        Return Me.GetEnumerator()
    End Function
#End Region

#Region "DirectOperations Logic"
    Friend Sub AddI(item As T)
        Me.InnerList.Add(item)
    End Sub

    Friend Sub ClearI()
        Me.InnerList.Clear()
    End Sub

    Friend Function RemoveI(item As T) As Boolean
        Return Me.InnerList.Remove(item)
    End Function
#End Region

    Public Sub Add(ByVal Item As T) Implements ICollection(Of T).Add
        Me._Master.AddC(Item, Me)
    End Sub

    Public Sub Clear() Implements ICollection(Of T).Clear
        Me._Master.ClearC(Me)
    End Sub

    Public Function Remove(ByVal Item As T) As Boolean Implements ICollection(Of T).Remove
        Return Me._Master.RemoveC(Item, Me)
    End Function

End Class

'Friend Class ComplementCollection(Of T)
'    Inherits Collection(Of T)
'    Implements INotifyCollectionChanged

'    Private ReadOnly InnerList As List(Of T)

'    Private Sub New(ByVal ObservingCollection As NotifyingCollection(Of T), ByVal MasterCollection As IEnumerable(Of T))
'        Me._ObservingCollection = ObservingCollection
'        Me._MasterCollection = MasterCollection
'        InnerList = New List(Of T)()
'    End Sub

'#Region "ObservingCollection Property"
'    Private ReadOnly _ObservingCollection As NotifyingCollection(Of T)

'    Public ReadOnly Property ObservingCollection As NotifyingCollection(Of T)
'        Get
'            Return Me._ObservingCollection
'        End Get
'    End Property
'#End Region

'#Region "MasterCollection Property"
'    Private ReadOnly _MasterCollection As IEnumerable(Of T)
'    Public Event CollectionChanged As NotifyCollectionChangedEventHandler Implements INotifyCollectionChanged.CollectionChanged

'    Public ReadOnly Property MasterCollection As IEnumerable(Of T)
'        Get
'            Return Me._MasterCollection
'        End Get
'    End Property
'#End Region

'    Private Sub OnObservingCollectionChanged(ByVal Sender As Object, ByVal E As NotifyCollectionChangedEventArgs)
'        Select Case E.Action
'            Case NotifyCollectionChangedAction.Add

'            Case NotifyCollectionChangedAction.Remove

'            Case NotifyCollectionChangedAction.Replace

'            Case NotifyCollectionChangedAction.Reset

'        End Select
'    End Sub

'    Protected Overrides Sub ClearItems()

'    End Sub

'    Protected Overrides Sub InsertItem(index As Integer, item As T)

'    End Sub

'    Protected Overridable Sub MoveItem(oldIndex As Integer, newIndex As Integer)

'    End Sub

'    Protected Overridable Sub OnCollectionChanged(e As NotifyCollectionChangedEventArgs)

'    End Sub

'    Protected Overrides Sub RemoveItem(index As Integer)

'    End Sub

'    Protected Overrides Sub SetItem(index As Integer, item As T)

'    End Sub

'End Class

File: Collections/ComplementCollection.cs
using System.Linq;
using System.Collections.Generic;
using System.Collections;
using System;
using System.Collections.Specialized;

namespace Ks.Common
{
    public class ComplementingCollectionMaster<T> : INotifyingCollection<T>
    {
        public ComplementingCollectionMaster()
        {
            this._Collection1 = new ChildComplementCollection<T>(this);
            this._Collection2 = new ChildComplementCollection<T>(this);
        }

        private readonly ChildComplementCollection<T> _Collection1;

        public INotifyingCollection<T> Collection1
        {
            get
            {
                return this._Collection1;
            }
        }

        private readonly ChildComplementCollection<T> _Collection2;

        public INotifyingCollection<T> Collection2
        {
            get
            {
                return this._Collection2;
            }
        }

        public event NotifyCollectionChangedEventHandler<T> CollectionChanged;
        event NotifyCollectionChangedEventHandler INotifyCollectionChanged.CollectionChanged
        {
            add => this.INotifyCollectionChanged_CollectionChanged += value;
            remove => this.INotifyCollectionChanged_CollectionChanged -= value;
        }
        private event NotifyCollectionChangedEventHandler INotifyCollectionChanged_CollectionChanged;

        protected virtual void OnCollectionChanged(NotifyCollectionChangedEventArgs<T> E)
        {
            CollectionChanged?.Invoke(this, E);
            INotifyCollectionChanged_CollectionChanged?.Invoke(this, E);
        }

        public int Count
        {
            get
            {
                return this._Collection1.Count + this._Collection2.Count;
            }
        }

        public bool IsReadOnly
        {
            get
            {
                return false;
            }
        }

        public void CopyTo(T[] array, int arrayIndex)
        {
            if ((array.Length - arrayIndex) < this.Count)
            {
                throw new ArgumentException("The number of elements in the source System.Collections.Generic.ICollection`1 is greater than the available space from arrayIndex to the end of the destination array.");
            }

            this._Collection1.CopyTo(array, arrayIndex);
            this._Collection2.CopyTo(array, arrayIndex + this._Collection1.Count);
        }

        public bool Contains(T item)
        {
            return this._Collection1.Contains(item) || this._Collection2.Contains(item);
        }

        public IEnumerator<T> GetEnumerator()
        {
            return this._Collection1.Concat(this._Collection2).GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        internal void AddC(T Item, ChildComplementCollection<T> C1, ChildComplementCollection<T> C2 = null)
        {
            if (C2 == null)
            {
                if (C1 == this._Collection1)
                {
                    C2 = this._Collection2;
                }
                else
                {
                    C2 = this._Collection1;
                }
            }

            if (!C2.RemoveI(Item))
            {
                throw new ArgumentException();
            }

            C1.AddI(Item);

            C1.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateAdd(Item));
            C2.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateRemove(Item));
        }

        internal void ClearC(ChildComplementCollection<T> C1, ChildComplementCollection<T> C2 = null)
        {
            IList<T> C1Clone;

            if (C2 == null)
            {
                if (C1 == this._Collection1)
                {
                    C2 = this._Collection2;
                }
                else
                {
                    C2 = this._Collection1;
                }
            }

            C1Clone = new T[C1.Count];
            C1.CopyTo(C1Clone);

            C1.ClearI();

            foreach (var I in C1Clone)
            {
                C2.AddI(I);
                C2.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateAdd(I));
            }

            C1.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateReset());
        }

        internal bool RemoveC(T Item, ChildComplementCollection<T> C1, ChildComplementCollection<T> C2 = null)
        {
            if (C2 == null)
            {
                if (C1 == this._Collection1)
                {
                    C2 = this._Collection2;
                }
                else
                {
                    C2 = this._Collection1;
                }
            }

            if (!C1.RemoveI(Item))
            {
                throw new ArgumentException();
            }

            C2.AddI(Item);

            C1.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateRemove(Item));
            C2.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateAdd(Item));

            return true;
        }

        public void Add(T item)
        {
            if (this._Collection1.Contains(item) || this._Collection2.Contains(item))
            {
                return;
            }

            this._Collection1.Add(item);
            this.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateAdd(item));
        }

        public void Clear()
        {
            this._Collection1.Clear();
            this._Collection2.Clear();
            this.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateReset());
        }

        public bool Remove(T item)
        {
            if (this._Collection1.Remove(item) || this._Collection2.Remove(item))
            {
                this.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateRemove(item));
                return true;
            }
            return false;
        }
    }

    internal class ChildComplementCollection<T> : INotifyingCollection<T>
    {
        internal ChildComplementCollection(ComplementingCollectionMaster<T> Master)
        {
            this.InnerList = new List<T>();
            this.Master = Master;
        }

        private readonly List<T> InnerList;

        public event NotifyCollectionChangedEventHandler<T> CollectionChanged;
        private event NotifyCollectionChangedEventHandler INotifyCollectionChanged_CollectionChanged;
        event NotifyCollectionChangedEventHandler INotifyCollectionChanged.CollectionChanged
        {
            add => this.INotifyCollectionChanged_CollectionChanged += value;
            remove => this.INotifyCollectionChanged_CollectionChanged -= value;
        }

        protected internal virtual void OnCollectionChanged(NotifyCollectionChangedEventArgs<T> E)
        {
            CollectionChanged?.Invoke(this, E);
            INotifyCollectionChanged_CollectionChanged?.Invoke(this, E);
        }

        public ComplementingCollectionMaster<T> Master { get; }

        public int Count
        {
            get
            {
                return this.InnerList.Count;
            }
        }

        public bool IsReadOnly
        {
            get
            {
                return false;
            }
        }

        public void CopyTo(T[] array, int arrayIndex)
        {
            this.InnerList.CopyTo(array, arrayIndex);
        }

        public bool Contains(T item)
        {
            return this.InnerList.Contains(item);
        }

        public List<T>.Enumerator GetEnumerator()
        {
            return this.InnerList.GetEnumerator();
        }

        IEnumerator<T> IEnumerable<T>.GetEnumerator()
        {
            return this.InnerList.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        internal void AddI(T item)
        {
            this.InnerList.Add(item);
        }

        internal void ClearI()
        {
            this.InnerList.Clear();
        }

        internal bool RemoveI(T item)
        {
            return this.InnerList.Remove(item);
        }

        public void Add(T Item)
        {
            this.Master.AddC(Item, this);
        }

        public void Clear()
        {
            this.Master.ClearC(this);
        }

        public bool Remove(T Item)
        {
            return this.Master.RemoveC(Item, this);
        }
    }
}
